include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+        values = values_double;
+      }
+
+      static void
+      jacobian(const types::tape_index         active_tape_index,
+               const unsigned int              n_dependent_variables,
+               const std::vector<scalar_type> &independent_variables,
+               FullMatrix<scalar_type> &       jacobian)
+      {
+        FullMatrix<double> jacobian_double(jacobian.m(), jacobian.n());
+        TapedDrivers<ADNumberType, double>::jacobian(active_tape_index,
+                                                     n_dependent_variables,
+                                                     vector_float_to_double(
+                                                       independent_variables),
+                                                     jacobian_double);
+        jacobian = jacobian_double;
+      }
+    };
+
+
+    // -------------   TapelessDrivers   -------------
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType, ScalarType, T>::initialize_global_environment(
+      const unsigned int)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    ScalarType
+    TapelessDrivers<ADNumberType, ScalarType, T>::value(
+      const std::vector<ADNumberType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+      return ScalarType(0.0);
+    }
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType, ScalarType, T>::gradient(
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType, ScalarType, T>::hessian(
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType, ScalarType, T>::values(
+      const std::vector<ADNumberType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template <typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapelessDrivers<ADNumberType, ScalarType, T>::jacobian(
+      const std::vector<ADNumberType> &,
+      const std::vector<ADNumberType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    namespace internal
+    {
+      // A dummy function to define the active dependent variable when using
+      // reverse-mode AD.
+      template <typename ADNumberType>
```

Comment:
this function and the ones below are internal, but they should probably still be documented using doxygen-style comments